teps/0121-unified-retries-for-runtime-objects.md

```diff
@@ -0,0 +1,433 @@
+surprising? Are there security implications for these implicit behaviors?
+-->
+
+### Flexibility
+WIP
+<!--
+https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility
+
+- Are there dependencies that need to be pulled in for this proposal to work? What support or maintenance would be
+required for these dependencies?
+- Are we coupling two or more Tekton projects in this proposal (e.g. coupling Pipelines to Chains)?
+- Are we coupling Tekton and other projects (e.g. Knative, Sigstore) in this proposal?
+- What is the impact of the coupling to operators e.g. maintenance & end-to-end testing?
+- Are there opinionated choices being made in this proposal? If so, are they necessary and can users extend it with
+their own choices?
+-->
+
+### Conformance
+WIP
+<!--
+https://github.com/tektoncd/community/blob/main/design-principles.md#conformance
+
+- Does this proposal require the user to understand how the Tekton API is implemented?
+- Does this proposal introduce additional Kubernetes concepts into the API? If so, is this necessary?
+- If the API is changing as a result of this proposal, what updates are needed to the
+[API spec](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md)?
+-->
+
+### User Experience
+WIP
+<!--
+(optional)
+
+Consideration about the user experience. Depending on the area of change,
+users may be Task and Pipeline editors, they may trigger TaskRuns and
+PipelineRuns or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+### Performance
+WIP
+<!--
+(optional)
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+- What impact does this change have on the start-up time and execution time
+of TaskRuns and PipelineRuns?
+- What impact does it have on the resource footprint of Tekton controllers
+as well as TaskRuns and PipelineRuns?
+-->
+
+### Risks and Mitigations
+WIP
+<!--
+What are the risks of this proposal and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Tekton ecosystem. Consider including folks that also work outside the WGs
+or subproject.
+- How will security be reviewed and by whom?
+- How will UX be reviewed and by whom?
+-->
+
+### Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+<!--
+What other approaches did you consider and why did you rule them out? These do
+not need to be as detailed as the proposal, but should include enough
+information to express the idea and why it was not acceptable.
+-->
+
+### Implements Retry in PipelineRun Controller
+
+Instead of letting runtime object controllers implement the retry functionality on their own, another option is **centralizing the implementation in PipelineRun Controller**. 
```

Comment:
This does simplify the implementation but it does not allow custom run controllers to implement specialized retries (as described in the first alternative here: https://github.com/tektoncd/community/blob/main/teps/0069-support-retries-for-custom-task-in-a-pipeline.md#alternatives)

Reply:
One alternative that's not discussed in the TEP is allowing customruns with specialized retry logic to configure that via the custom task spec. The main reason I prefer having retries implemented by the pipelinerun is that I don't think `retries` belong in `customRun.spec.retries` (as opposed to `customRun.spec.customSpec`), since not every custom task has a concept of retries.

Reply:
Wouldn't that mean that for custom tasks that have a concept of specialized retries, a user would have to choose between which retry field to specify?
I think we should be clearer around what the behavior should be if a custom task does not specify a retry - maybe it should fail the run or provide a warning?

Reply:
Thanks for all of your comments @dibyom and @lbernick.

> it does not allow custom run controllers to implement specialized retries (as described in the first alternative here

Thanks for the pointer @dibyom , added this information to the first alternative.

> One alternative that's not discussed in the TEP is allowing customruns with specialized retry logic to configure that via the custom task spec. 

Thanks for the comment @lbernick, as [discussed in the API WG](https://docs.google.com/document/d/17PodAxG8hV351fBhSu7Y_OIPhGTVgj6OJ2lPphYYRpU/edit#bookmark=id.hwc5acp8tkm), we expect every custom run controller to support retries. Put this  as the sencond alternatives of this proposal.

> it does not allow custom run controllers to implement specialized retries.

Thanks for the pointer Dibyo, indeed, added this information to the alternatives.

> Wouldn't that mean that for custom tasks that have a concept of specialized retries, a user would have to choose between which retry field to specify?

Put it as a concern of the second alternative of this TEP, this is truely not quite user-friendly.